Re-apply "Make all CF types Equatable and Hashable."

jrose-apple opened the pull request:
This re-applies #4394, reverted in #4435, now that John has fixed CF casting logic in the optimizer. The round-trip casting test is also passing now.

https://bugs.swift.org/browse/SR-2388

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@DougGregor, @parkera, this is technically a source-breaking change since someone may have manually added Hashable or Equatable to a CF type. Then again, that's both a type and protocol they don't own, which isn't really safe because *anyone* could do it ("what if two people did this?").  Do you think we can commit this or do we have to be more circumspect?

lattner commented:
This seems like acceptable risk to me.

karwa commented on test/Interpreter/SDK/cf_without_foundation.swift:
```diff
@@ -0,0 +1,61 @@
+// RUN: %target-run-simple-swift | %FileCheck %s
+
+// REQUIRES: executable_test
+// REQUIRES: objc_interop
```
Technically only requires a platform where CF is part of the SDK, doesn't it?

What do you think about moving/copying this test in to Foundation and running it on all platforms?

jrose-apple commented:
I've found at least one case on GitHub where someone is doing exactly this on CFString, so it's definitely a source-breaking change. May have to wait until Swift 4.

thomasbarwanitz reviewed (commented):
Thank you

thomasbarwanitz reviewed (commented):
Thhhhxzzzzzzzzzz

jrose-apple commented:
Resurrecting this. It's almost certainly breaking, and we don't have a way to conditionalize it on Swift 3/4, but it is an improvement. What do people think?

dabrahams commented:
Even if this breaks somebody, it's really simple to fix.  +1

DougGregor commented:
I'm definitely +1 on doing this

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
I still want to do a full test on Apple platforms but there's an error in one test right now.

@swift-ci Please smoke test

jrose-apple committed:
Re-apply "Make all CF types Equatable and Hashable."

Like NSObject, CFType has primitive operations CFEqual and CFHash,
so Swift should allow those types to show up in Hashable positions
(like dictionaries). The most general way to do this was to
introduce a new protocol, _CFObject, and then have the importer
automatically make all CF types conform to it.

This did require one additional change: the == implementation that
calls through to CFEqual is in a new CoreFoundation overlay, but the
conformance is in the underlying Clang module. Therefore, operator
lookup for conformances has been changed to look in the overlay for
an imported declaration (if there is one).

This re-applies 361ab6245405c1ab8b9831748395b1055b9df595, reverted in
f50b1e73dc3144fc3abb1ff540d06a1fbb5853cf, after a /very/ long interval
where we decided if it was worth breaking people who've added these
conformances on their own. Since the workaround isn't too difficult---
use `#if swift(>=3.2)` to guard the extension introducing the
conformance---it was deemed acceptable.

https://bugs.swift.org/browse/SR-2388

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@akyrtzi, we're now running into the old issue where `-include` doesn't actually count as having included something (rdar://problem/22083824). Are you okay with the change to test/IDE/print_clang_header.swift to kick that can down the road a little further?

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please smoke test
